File: tests/typesafety/typescript/dispatch_untyped.ts
// expect: passes

type Invoice = { readonly number: string; readonly status: string };

function send(invoice: Invoice): Invoice {
  return { number: invoice.number, status: "sent" };
}

function archive(invoiceNumber: string): string {
  return `archived ${invoiceNumber}`;
}

const handlers: Record<string, (...args: any[]) => any> = {
  draft: send,
  sent: archive,
};

export function advance(invoice: Invoice): any {
  return handlers[invoice.status]?.(invoice);
}

File: tests/typesafety/python/dispatch_untyped.py
# expect: passes
# title: A table of handlers typed Any takes a handler that cannot run
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class Invoice:
    number: str
    status: str


def send(invoice: Invoice) -> Invoice:
    return Invoice(invoice.number, "sent")


def archive(number: str) -> str:
    return f"archived {number}"


HANDLERS: dict[str, Callable[..., Any]] = {
    "draft": send,
    "sent": archive,
}


def advance(invoice: Invoice) -> Any:
    return HANDLERS[invoice.status](invoice)
